lib/puma/runner.rb

```diff
@@ -75,7 +75,11 @@ def start_control
       control = Puma::Server.new app, nil,
         { min_threads: 0, max_threads: 1, queue_requests: false, log_writer: @log_writer }
 
-      control.binder.parse [str], nil, 'Starting control server'
+      begin
+        control.binder.parse [str], nil, 'Starting control server'
+      rescue Errno::EADDRINUSE => e
+        raise "Error: Control server address '#{str}' is already in use. You're trying to start a control server but the control server port is taken."
+      end
```

Comment:
Should we also re-raise the actual error? 

We could rescue all errors, add our helpful message and re-raise. I think that would cover https://github.com/puma/puma/issues/2894 too (seems like the exception raised on Windows is a different one)

Reply:
@dentarg are you referring to something like this?

```ruby
begin
  control.binder.parse [str], nil, 'Starting control server'
rescue Errno::EADDRINUSE, Errno::EACCES => e
  raise e, "Error: Control server address '#{str}' is already in use. You're trying to start a control server but the control server port is taken."
end
```

Also do you think there is some value in creating a descriptive exception class as well? 

(Perhaps `StandardError` isn't the right exception class to inherit from?)
```ruby
class ControlServerAddressPortConflict < StandardError; end
```

```ruby
begin
  control.binder.parse [str], nil, 'Starting control server'
rescue Errno::EADDRINUSE, Errno::EACCES => e
  raise ControlServerAddressPortConflict, "Error: Control server address '#{str}' is already in use. You're trying to start a control server but the control server port is taken."
end
```

Reply:
Now it should cover #2894 too i think.

Reply:
Actually, after thinking about this I'm not so sure that it would handle #2894. Not that it couldn't but I would imagine the error message would at least need to be different. Sounds like @MSP-Greg started a branch for this last month but hasn't posted yet. I'd be interested in seeing what that looks like

Reply:
yeah, might need to make the error msg a bit generic in that case.

Reply:
Also, i think 
`class ControlServerAddressPortConflict < StandardError; end`
The error should be abstracted out in case we are raising Control server port issues in multiple (more than 1?) places. Otherwise doesn't make much sense IMO. Leme know what you guys think

Reply:
I see where you're coming from with the abstraction comment, but I think it's a good idea because it will lead to a more helpful exception message stack trace. Even if it isn't used other places, I think that wrapping the exception in more domain specific terminology can help increase clarity imo 🙂. But it's your PR so use your discretion

Reply:
For me, Errno::EADDRINUSE, Errno::EACCES is clear enough 🤷 

I think we should rescue `Errno::EADDRINUSE, Errno::EACCES` here, as StandardError is too broad and may display this message in confusing circumstances.

Reply:
And as dentarg said, we should re-raise.

Reply:
Fixed, re-raised the original error too.